File: adafruit-imu-bno055/python/imu-bno055.py
#!/usr/bin/env python3
"""
Copyright (c) 2025, BlackBerry Limited. All rights reserved.
 
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at
 
     http://www.apache.org/licenses/LICENSE-2.0
 
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
 
@file imu-bno055.py
@brief Adafruit BNO055 9-DOF IMU Example for QNX.
 
This example demonstrates how to interface with the Adafruit BNO055 Absolute
Orientation Sensor over I2C on a QNX system using Python's smbus module. It:
  - Initializes the sensor and verifies its chip ID (expected value: 0xA0).
  - Implements a calibration system:
    - Attempts to load existing calibration data from a file.
    - If no file exists or if `--calibrate` is passed as an argument, it
      initiates a guided, manual calibration process.
    - Saves the new calibration data to a file for future use.
  - Sets the sensor to NDOF (Nine Degrees of Freedom) fusion mode.
  - Continuously reads and prints all fused and raw sensor data.
 
Wiring (Raspberry Pi header mapping):
  - VIN: Connect to PIN 1 (3.3V Power).
  - GND: Connect to PIN 6 (Ground).
  - SCL: Connect to PIN 5 (GPIO 3).
  - SDA: Connect to PIN 3 (GPIO 2).
"""
 
import smbus
import time
import sys
import argparse
import os
 
## --------------------------------------------------------------------------
## Definitions
## --------------------------------------------------------------------------
BNO055_I2C_BUS          = 1
BNO055_ADDRESS          = 0x28
CALIBRATION_FILENAME    = "bno055_calib.dat"
 
# BNO055 Register Map
BNO055_CHIP_ID_ADDR          = 0x00
BNO055_OPR_MODE_ADDR         = 0x3D
BNO055_SYS_TRIGGER_ADDR      = 0x3F
BNO055_CALIB_STAT_ADDR       = 0x35
BNO055_ACC_OFFSET_X_LSB_ADDR = 0x55
CALIBRATION_DATA_SIZE        = 22
 
# Data Registers
BNO055_ACC_DATA_X_LSB_ADDR  = 0x08
BNO055_MAG_DATA_X_LSB_ADDR  = 0x0E
BNO055_GYR_DATA_X_LSB_ADDR  = 0x14
BNO055_EUL_HEADING_LSB_ADDR = 0x1A
BNO055_QUA_DATA_W_LSB_ADDR  = 0x20
BNO055_LIA_DATA_X_LSB_ADDR  = 0x28
BNO055_GRV_DATA_X_LSB_ADDR  = 0x2E
 
# BNO055 Operation Modes
BNO055_OPERATION_MODE_CONFIG = 0x00
BNO055_OPERATION_MODE_NDOF   = 0x0C

# Data Scaling Divisors (LSB to physical units)
BNO055_ACCEL_DIVISOR    = 100.0
BNO055_MAG_DIVISOR      = 16.0
BNO055_GYRO_DIVISOR     = 16.0
BNO055_EULER_DIVISOR    = 16.0
BNO055_QUAT_SCALE       = 1.0 / (1 << 14)
 
## --------------------------------------------------------------------------
## Create SMBus Object
## --------------------------------------------------------------------------
bus = smbus.SMBus(BNO055_I2C_BUS)
 
## --------------------------------------------------------------------------
## Helper Functions for Register Read/Write
## --------------------------------------------------------------------------
def _signed_int16(val: int) -> int:
    """Converts a 16-bit unsigned value to a signed integer."""
    return val - 65536 if val & 0x8000 else val
 
def bno055_set_mode(mode: int):
    """
    @brief Sets the operation mode of the BNO055.
    @param mode The desired operation mode (e.g., BNO055_OPERATION_MODE_CONFIG).
    """
    try:
        bus.write_byte_data(BNO055_ADDRESS, BNO055_OPR_MODE_ADDR, mode)
        time.sleep(0.02) # Wait for mode switch (datasheet recommends >19ms)
    except OSError as e:
        print(f"ERROR: Could not set mode: {e}", file=sys.stderr)
        sys.exit(1)
 
## --------------------------------------------------------------------------
## Sensor Initialization
## --------------------------------------------------------------------------
def bno055_init() -> int:
    """
    @brief Initializes the BNO055 sensor.
    
    This function performs the following:
      - Verifies the sensor's CHIP_ID.
      - Resets the sensor.
    
    @return int 0 on success, or a nonzero error code.
    """
    try:
        chip_id = bus.read_byte_data(BNO055_ADDRESS, BNO055_CHIP_ID_ADDR)
        if chip_id != 0xA0:
            print(f"ERROR: BNO055 not found at I2C address 0x{BNO055_ADDRESS:02X}. Chip ID: 0x{chip_id:02X}", file=sys.stderr)
            return 1
        print(f"BNO055 Chip ID: 0x{chip_id:02X}")
        
        # Reset the sensor
        bno055_set_mode(BNO055_OPERATION_MODE_CONFIG)
        bus.write_byte_data(BNO055_ADDRESS, BNO055_SYS_TRIGGER_ADDR, 0x20)
        time.sleep(0.7) # Wait for reset to complete
        return 0
    except OSError as e:
        print(f"ERROR: During sensor initialization: {e}", file=sys.stderr)
        return 1
 
## --------------------------------------------------------------------------
## Sensor Data Reading
## --------------------------------------------------------------------------
def bno055_read_vector(reg: int, divisor: float) -> tuple:
    """
    @brief Reads a 3-axis vector from the BNO055.
    @param reg The starting register address for the 3-axis data (6 bytes).
    @param divisor The scaling factor to convert the raw value to physical units.
    @return tuple A tuple (x, y, z) of the scaled vector data.
    """
    try:
        data = bus.read_i2c_block_data(BNO055_ADDRESS, reg, 6)
        x = _signed_int16(data[0] | (data[1] << 8)) / divisor
        y = _signed_int16(data[2] | (data[3] << 8)) / divisor
        z = _signed_int16(data[4] | (data[5] << 8)) / divisor
        return (x, y, z)
    except OSError:
        return (0, 0, 0)
 
def bno055_read_quaternion() -> tuple:
    """
    @brief Reads the quaternion data from the BNO055.
    @return tuple A tuple (w, x, y, z) of the quaternion data.
    """
    scale = 1.0 / (1 << 14)
    try:
        data = bus.read_i2c_block_data(BNO055_ADDRESS, BNO055_QUA_DATA_W_LSB_ADDR, 8)
        w = _signed_int16(data[0] | (data[1] << 8)) * scale
        x = _signed_int16(data[2] | (data[3] << 8)) * scale
        y = _signed_int16(data[4] | (data[5] << 8)) * scale
        z = _signed_int16(data[6] | (data[7] << 8)) * scale
        return (w, x, y, z)
    except OSError:
        return (0, 0, 0, 0)
 
def bno055_read_and_print_all_data():
    """
    @brief Reads all available sensor data and prints it to the console.
    """
    euler = bno055_read_vector(BNO055_EUL_HEADING_LSB_ADDR, BNO055_EULER_DIVISOR)
    accel = bno055_read_vector(BNO055_ACC_DATA_X_LSB_ADDR, BNO055_ACCEL_DIVISOR)
    mag = bno055_read_vector(BNO055_MAG_DATA_X_LSB_ADDR, BNO055_MAG_DIVISOR)
    gyro = bno055_read_vector(BNO055_GYR_DATA_X_LSB_ADDR, BNO055_GYRO_DIVISOR)
    lia = bno055_read_vector(BNO055_LIA_DATA_X_LSB_ADDR, BNO055_ACCEL_DIVISOR)
    grv = bno055_read_vector(BNO055_GRV_DATA_X_LSB_ADDR, BNO055_ACCEL_DIVISOR)
    quat = bno055_read_quaternion()
 
    print("--------------------------------------------------")
    print(f"Euler (H,R,P):        {euler[0]:7.2f}, {euler[1]:7.2f}, {euler[2]:7.2f} deg")
    print(f"Quaternion (W,X,Y,Z): {quat[0]:7.2f}, {quat[1]:7.2f}, {quat[2]:7.2f}, {quat[3]:7.2f}")
    print(f"Accelerometer (m/s^2):  X={accel[0]:7.2f}, Y={accel[1]:7.2f}, Z={accel[2]:7.2f}")
    print(f"Linear Accel (m/s^2):   X={lia[0]:7.2f}, Y={lia[1]:7.2f}, Z={lia[2]:7.2f}")
    print(f"Gravity Vector (m/s^2): X={grv[0]:7.2f}, Y={grv[1]:7.2f}, Z={grv[2]:7.2f}")
    print(f"Gyroscope (dps):        X={gyro[0]:7.2f}, Y={gyro[1]:7.2f}, Z={gyro[2]:7.2f}")
    print(f"Magnetometer (uT):      X={mag[0]:7.2f}, Y={mag[1]:7.2f}, Z={mag[2]:7.2f}")
 
## --------------------------------------------------------------------------
## Calibration Routine
## --------------------------------------------------------------------------
def bno055_perform_manual_calibration():
    """
    @brief Guides the user through a manual calibration process.
    """
    print("\n--- Starting Manual Calibration ---")
    print("Please perform the calibration dance until all components are 3:")
    print("1. Gyroscope (GYR): Place on a flat, stable surface.")
    print("2. Accelerometer (ACC): Slowly move to ~6 different stable orientations.")
    print("3. Magnetometer (MAG): Slowly make a large figure-8 motion.\n")
 
    while True:
        try:
            calib_stat = bus.read_byte_data(BNO055_ADDRESS, BNO055_CALIB_STAT_ADDR)
            sys_calib = (calib_stat >> 6) & 0x03
            gyro_calib = (calib_stat >> 4) & 0x03
            accel_calib = (calib_stat >> 2) & 0x03
            mag_calib = calib_stat & 0x03
 
            print(f"Calibration Status: SYS={sys_calib}, GYR={gyro_calib}, ACC={accel_calib}, MAG={mag_calib}", end='\r', flush=True)
 
            if gyro_calib == 3 and accel_calib == 3 and mag_calib == 3 and sys_calib == 3:
                print("\nFull System Calibration Complete!")
                break
            time.sleep(0.1)
        except OSError as e:
            print(f"\nError reading calibration status: {e}", file=sys.stderr)
            time.sleep(1)
 
def bno055_save_calibration(filename: str):
    """
    @brief Saves the sensor's calibration data to a file.
    @param filename The name of the file to save the data to.
    """
    print(f"Saving calibration data to {filename}...")
    bno055_set_mode(BNO055_OPERATION_MODE_CONFIG)
    try:
        calib_data = bus.read_i2c_block_data(BNO055_ADDRESS, BNO055_ACC_OFFSET_X_LSB_ADDR, CALIBRATION_DATA_SIZE)
        with open(filename, 'wb') as f:
            f.write(bytearray(calib_data))
        print("Successfully saved calibration data.")
    except (OSError, IOError) as e:
        print(f"Error saving calibration data: {e}", file=sys.stderr)
 
def bno055_load_calibration(filename: str) -> bool:
    """
    @brief Loads calibration data from a file and writes it to the sensor.
    @param filename The name of the file to load data from.
    @return bool True if loading was successful, False otherwise.
    """
    if not os.path.exists(filename):
        print("INFO: Calibration file not found. Starting manual calibration.")
        return False
    
    print(f"Loading calibration data from {filename}...")
    try:
        with open(filename, 'rb') as f:
            calib_data = list(f.read(CALIBRATION_DATA_SIZE))
        
        bno055_set_mode(BNO055_OPERATION_MODE_CONFIG)
        bus.write_i2c_block_data(BNO055_ADDRESS, BNO055_ACC_OFFSET_X_LSB_ADDR, calib_data)
        print("Successfully loaded calibration data.")
        return True
    except (OSError, IOError) as e:
        print(f"Error loading calibration data: {e}", file=sys.stderr)
        return False
 
## --------------------------------------------------------------------------
## Main Application Loop
## --------------------------------------------------------------------------
def main():
    """
    @brief Main function.
    
    The main loop performs the following:
      - Initializes the sensor.
      - Handles calibration loading or execution.
      - Enters a loop to continuously read and display sensor data.
    """
    parser = argparse.ArgumentParser(description="BNO055 IMU reader for QNX.")
    parser.add_argument('--calibrate', action='store_true', help='Force manual recalibration.')
    args = parser.parse_args()
 
    print("=== Adafruit BNO055 Demo ===")
 
    rc = bno055_init()
    if rc != 0:
        print("ERROR: bno055_init() failed.", file=sys.stderr)
        sys.exit(1)
 
   # Load calibration or perform manual calibration
    if args.calibrate or not bno055_load_calibration(CALIBRATION_FILENAME):
        bno055_set_mode(BNO055_OPERATION_MODE_NDOF)
        bno055_perform_manual_calibration()
        bno055_save_calibration(CALIBRATION_FILENAME)
 
    # Set final operation mode for data reading
    print("Setting final operation mode to NDOF...")
    bno055_set_mode(BNO055_OPERATION_MODE_NDOF)
    
    print("\n--- Starting Continuous Data Readout (Press Ctrl+C to exit) ---")
    while True:
        bno055_read_and_print_all_data()
        time.sleep(0.1) # ~10Hz
 
if __name__ == '__main__':
    try:
        main()
    except KeyboardInterrupt:
        print("\nExiting BNO055 demo.")
        # The smbus object's destructor will handle closing the bus connection.
        sys.exit(0)
    except OSError as e:
        print(f"\nA fatal bus error occurred: {e}", file=sys.stderr)
        sys.exit(1)
 

File: adafruit-imu-bno055/c/src/imu-bno055.c
/**
 * Copyright (c) 2025, BlackBerry Limited. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * @file imu-bno055.c
 * @brief Adafruit BNO055 9-DOF IMU Example for QNX.
 *
 * This example demonstrates how to interface with the Adafruit BNO055 Absolute
 *   Orientation Sensor over I2C on a QNX system. It:
 * - Initializes the sensor and verifies its chip ID (expected value: 0xA0).
 * - Implements a calibration system:
 * - Attempts to load existing calibration data from a file.
 * - If no file exists or if `--calibrate` is passed as an argument, it
 *   initiates a guided, manual calibration process.
 * - Saves the new calibration data to a file for future use.
 * - Sets the sensor to NDOF (Nine Degrees of Freedom) fusion mode.
 * - Continuously reads and prints all fused and raw sensor data, including:
 * - Euler Angles (Heading, Roll, Pitch)
 * - Quaternions
 * - Accelerometer (raw, linear, and gravity vector)
 * - Gyroscope
 * - Magnetometer
 *
 * Wiring: (Raspberry Pi header mapping):
 * - VIN: Connect to PIN 1 (3.3V Power).
 * - GND: Connect to PIN 6 (Ground).
 * - SCL: Connect to PIN 5 (GPIO 3).
 * - SDA: Connect to PIN 3 (GPIO 2).
 * - RST: Connect to PIN 13 (GPIO 27) - Note: Reset logic is not used in this example.
 * - INT: Connect to PIN 15 (GPIO 22) - Note: Interrupts are not used in this example.
 *
 * Data Scaling:
 * - Accelerometer, Linear Acceleration, Gravity: 1m/s² = 100 LSB
 * - Magnetometer: 1uT = 16 LSB
 * - Gyroscope: 1°/s = 16 LSB
 * - Euler Angles: 1° = 16 LSB
 * - Quaternion: 1 LSB = 2^-14
 *
 * - Compile:
 *   qcc -Vgcc_ntoaarch64le -o imu-bno055 imu-bno055.c
 */
 
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <stdint.h>
#include <string.h>
#include "rpi_i2c.h" 
 
/* --------------------------------------------------------------------------
 * BNO055 Configuration and Definitions
 * ------------------------------------------------------------------------*/
 
// --- I2C Configuration ---
#define BNO055_ADDRESS_A        (0x28)             ///< Default I2C address for the BNO055
#define BNO055_I2C_BUS          1                  ///< I2C bus number
#define CALIBRATION_FILENAME    "bno055_calib.dat" ///< File to store calibration data
 
// --- BNO055 Register Map ---
#define BNO055_CHIP_ID_ADDR          (0x00)  ///< Chip ID Register; expected value: 0xA0
#define BNO055_OPR_MODE_ADDR         (0x3D)  ///< Operation Mode Register
#define BNO055_SYS_TRIGGER_ADDR      (0x3F)  ///< System Trigger Register (for reset)
#define BNO055_CALIB_STAT_ADDR       (0x35)  ///< Calibration Status Register
#define BNO055_ACC_OFFSET_X_LSB_ADDR (0x55)  ///< Start of calibration data registers
#define CALIBRATION_DATA_SIZE        (22)    ///< Size of the calibration data block (22 bytes)
 
// --- Data Registers ---
#define BNO055_ACC_DATA_X_LSB_ADDR  (0x08)  ///< Accelerometer data start
#define BNO055_MAG_DATA_X_LSB_ADDR  (0x0E)  ///< Magnetometer data start
#define BNO055_GYR_DATA_X_LSB_ADDR  (0x14)  ///< Gyroscope data start
#define BNO055_EUL_HEADING_LSB_ADDR (0x1A)  ///< Euler angle data start
#define BNO055_QUA_DATA_W_LSB_ADDR  (0x20)  ///< Quaternion data start
#define BNO055_LIA_DATA_X_LSB_ADDR  (0x28)  ///< Linear Acceleration data start
#define BNO055_GRV_DATA_X_LSB_ADDR  (0x2E)  ///< Gravity Vector data start
 
// --- BNO055 Operation Modes ---
#define BNO055_OPERATION_MODE_CONFIG (0x00) ///< Configuration mode (for setup)
#define BNO055_OPERATION_MODE_NDOF   (0x0C) ///< 9-DOF absolute orientation fusion mode

// --- Data Scaling Divisors (LSB to physical units) ---
#define BNO055_ACCEL_DIVISOR    100.0  // 1m/s² = 100 LSB
#define BNO055_MAG_DIVISOR      16.0   // 1uT = 16 LSB
#define BNO055_GYRO_DIVISOR     16.0   // 1dps = 16 LSB
#define BNO055_EULER_DIVISOR    16.0   // 1° = 16 LSB
#define BNO055_QUAT_SCALE       (1.0 / (1 << 14)) // Scale for quaternion values
 
/* --------------------------------------------------------------------------
 * Data Structures
 * ------------------------------------------------------------------------*/
 
/**
 * @brief Stores a 3-axis vector (e.g., for accelerometer, gyroscope).
 */
typedef struct {
    double x, y, z;
} bno055_vector_t;
 
/**
 * @brief Stores a 4-component quaternion.
 */
typedef struct {
    double w, x, y, z;
} bno055_quaternion_t;

/* --------------------------------------------------------------------------
* Function Prototypes
* ------------------------------------------------------------------------*/
 
int bno055_set_mode(uint8_t mode);
int bno055_save_calibration(const char* filename);
int bno055_load_calibration(const char* filename);
void bno055_perform_manual_calibration();
void read_all_sensor_data();
int bno055_read_vector(uint8_t reg, bno055_vector_t *vec, double divisor);
int bno055_read_quaternion(bno055_quaternion_t *quat);

/* --------------------------------------------------------------------------
 * Function Implementations
 * ------------------------------------------------------------------------*/
 
/**
 * @brief Reads all available sensor data and prints it to the console.
 */
void read_all_sensor_data() {
    bno055_vector_t euler, accel, mag, gyro, lia, grv;
    bno055_quaternion_t quat;
 
    // Read all sensor vectors using the defined constants
    bno055_read_vector(BNO055_EUL_HEADING_LSB_ADDR, &euler, BNO055_EULER_DIVISOR);
    bno055_read_vector(BNO055_ACC_DATA_X_LSB_ADDR, &accel, BNO055_ACCEL_DIVISOR);
    bno055_read_vector(BNO055_MAG_DATA_X_LSB_ADDR, &mag, BNO055_MAG_DIVISOR);
    bno055_read_vector(BNO055_GYR_DATA_X_LSB_ADDR, &gyro, BNO055_GYRO_DIVISOR);
    bno055_read_vector(BNO055_LIA_DATA_X_LSB_ADDR, &lia, BNO055_ACCEL_DIVISOR); // Linear Accel shares divisor with Accel
    bno055_read_vector(BNO055_GRV_DATA_X_LSB_ADDR, &grv, BNO055_ACCEL_DIVISOR); // Gravity shares divisor with Accel
    bno055_read_quaternion(&quat);
 
    printf("--------------------------------------------------\n");
    printf("Euler (H,R,P):        %7.2f, %7.2f, %7.2f deg\n", euler.x, euler.y, euler.z);
    printf("Quaternion (W,X,Y,Z): %7.2f, %7.2f, %7.2f, %7.2f\n", quat.w, quat.x, quat.y, quat.z);
    printf("Accelerometer (m/s^2):  X=%7.2f, Y=%7.2f, Z=%7.2f\n", accel.x, accel.y, accel.z);
    printf("Linear Accel (m/s^2):   X=%7.2f, Y=%7.2f, Z=%7.2f\n", lia.x, lia.y, lia.z);
    printf("Gravity Vector (m/s^2): X=%7.2f, Y=%7.2f, Z=%7.2f\n", grv.x, grv.y, grv.z);
    printf("Gyroscope (dps):        X=%7.2f, Y=%7.2f, Z=%7.2f\n", gyro.x, gyro.y, gyro.z);
    printf("Magnetometer (uT):      X=%7.2f, Y=%7.2f, Z=%7.2f\n", mag.x, mag.y, mag.z);
}
 
/**
 * @brief Reads a 3-axis vector from the BNO055.
 *
 * @param reg The starting register address for the 3-axis data (6 bytes).
 * @param vec Pointer to a `bno055_vector_t` struct to store the result.
 * @param divisor The scaling factor to convert the raw integer value to physical units.
 * @return int 0 on success, -1 on I2C read error.
 */
int bno055_read_vector(uint8_t reg, bno055_vector_t *vec, double divisor) {
    uint8_t buffer[6];
    if (smbus_read_block_data(BNO055_I2C_BUS, BNO055_ADDRESS_A, reg, buffer, 6) != 0) {
        return -1;
    }
 
    int16_t x = (int16_t)(((uint16_t)buffer[1] << 8) | (uint16_t)buffer[0]);
    int16_t y = (int16_t)(((uint16_t)buffer[3] << 8) | (uint16_t)buffer[2]);
    int16_t z = (int16_t)(((uint16_t)buffer[5] << 8) | (uint16_t)buffer[4]);
 
    vec->x = (double)x / divisor;
    vec->y = (double)y / divisor;
    vec->z = (double)z / divisor;
 
    return 0;
}
 
/**
 * @brief Reads the quaternion data from the BNO055.
 *
 * @param quat Pointer to a `bno055_quaternion_t` struct to store the result.
 * @return int 0 on success, -1 on I2C read error.
 */
int bno055_read_quaternion(bno055_quaternion_t *quat) {
    uint8_t buffer[8];
    if (smbus_read_block_data(BNO055_I2C_BUS, BNO055_ADDRESS_A, BNO055_QUA_DATA_W_LSB_ADDR, buffer, 8) != 0) {
        return -1;
    }
 
    double scale = (1.0 / (1 << 14)); // 1.0 / 16384.0
 
    quat->w = (double)((int16_t)(((uint16_t)buffer[1] << 8) | (uint16_t)buffer[0])) * scale;
    quat->x = (double)((int16_t)(((uint16_t)buffer[3] << 8) | (uint16_t)buffer[2])) * scale;
    quat->y = (double)((int16_t)(((uint16_t)buffer[5] << 8) | (uint16_t)buffer[4])) * scale;
    quat->z = (double)((int16_t)(((uint16_t)buffer[7] << 8) | (uint16_t)buffer[6])) * scale;
 
    return 0;
}
 
/**
 * @brief Sets the operation mode of the BNO055.
 *
 * @param mode The desired operation mode (e.g., BNO055_OPERATION_MODE_CONFIG).
 * @return int 0 on success, -1 on I2C write error.
 */
int bno055_set_mode(uint8_t mode) {
    if (smbus_write_byte_data(BNO055_I2C_BUS, BNO055_ADDRESS_A, BNO055_OPR_MODE_ADDR, mode) != 0) {
        perror("Failed to set operation mode");
        return -1;
    }
    return 0;
}
 
/**
 * @brief Guides the user through a manual calibration process.
 *
 * This function continuously polls the calibration status register and prints
 * the status for the system, gyroscope, accelerometer, and magnetometer.
 * It waits until the gyroscope, accelerometer, and magnetometer all report
 * full calibration (status 3) before completing.
 */
void bno055_perform_manual_calibration() {
    uint8_t calib_stat = 0;
    printf("\n--- Starting Manual Calibration ---\n");
    printf("Please perform the calibration dance until all components are 3:\n");
    printf("1. Gyroscope (GYR): Place on a flat, stable surface.\n");
    printf("2. Accelerometer (ACC): Slowly move to ~6 different stable orientations.\n");
    printf("3. Magnetometer (MAG): Slowly make a large figure-8 motion.\n\n");
 
    while (1) {
        smbus_read_byte_data(BNO055_I2C_BUS, BNO055_ADDRESS_A, BNO055_CALIB_STAT_ADDR, &calib_stat);
 
        uint8_t sys_calib = (calib_stat >> 6) & 0x03;
        uint8_t gyro_calib = (calib_stat >> 4) & 0x03;
        uint8_t accel_calib = (calib_stat >> 2) & 0x03;
        uint8_t mag_calib = calib_stat & 0x03;
 
        printf("Calibration Status: SYS=%d, GYR=%d, ACC=%d, MAG=%d\r", sys_calib, gyro_calib, accel_calib, mag_calib);
        fflush(stdout);
 
        // Wait until individual sensors are fully calibrated.
        if (gyro_calib == 3 && accel_calib == 3 && mag_calib == 3) {
            // Once individual sensors are good, the system fusion (SYS) will also stabilize.
            if (sys_calib == 3) {
                 printf("\nFull System Calibration Complete!\n");
                 break;
            }
        }
        usleep(100000);
    }
}
 
/**
 * @brief Saves the sensor's calibration data to a file.
 *
 * This function switches the sensor to CONFIG mode, reads the 22-byte
 * calibration offset and radius data, and writes it to the specified file.
 *
 * @param filename The name of the file to save the data to.
 * @return int 0 on success, -1 on failure.
 */
int bno055_save_calibration(const char* filename) {
    uint8_t calib_data[CALIBRATION_DATA_SIZE];
 
    // Switch to CONFIG mode to read calibration data
    bno055_set_mode(BNO055_OPERATION_MODE_CONFIG);
    usleep(25000);
 
    if (smbus_read_block_data(BNO055_I2C_BUS, BNO055_ADDRESS_A, BNO055_ACC_OFFSET_X_LSB_ADDR, calib_data, CALIBRATION_DATA_SIZE) != 0) {
        perror("Failed to read calibration data from sensor");
        return -1;
    }
 
    FILE* fp = fopen(filename, "w");
    if (!fp) {
        perror("Failed to open calibration file for writing");
        return -1;
    }
 
    fwrite(calib_data, 1, CALIBRATION_DATA_SIZE, fp);
    fclose(fp);
    printf("Successfully saved calibration data to %s\n", filename);
    return 0;
}
 
/**
 * @brief Loads calibration data from a file and writes it to the sensor.
 *
 * This function reads the 22-byte calibration data from the specified file,
 * switches the sensor to CONFIG mode, and writes the data to the sensor's
 * offset and radius registers.
 *
 * @param filename The name of the file to load data from.
 * @return int 0 on success, -1 on failure (e.g., file not found).
 */
int bno055_load_calibration(const char* filename) {
    uint8_t calib_data[CALIBRATION_DATA_SIZE];
 
    FILE* fp = fopen(filename, "r");
    if (!fp) {
        printf("INFO: Calibration file not found. Starting manual calibration.\n");
        return -1;
    }
 
    fread(calib_data, 1, CALIBRATION_DATA_SIZE, fp);
    fclose(fp);
 
    // Switch to CONFIG mode to write calibration data
    bno055_set_mode(BNO055_OPERATION_MODE_CONFIG);
    usleep(25000);
 
    if (smbus_write_block_data(BNO055_I2C_BUS, BNO055_ADDRESS_A, BNO055_ACC_OFFSET_X_LSB_ADDR, calib_data, CALIBRATION_DATA_SIZE) != 0) {
        perror("Failed to write calibration data to sensor");
        return -1;
    }
    printf("Successfully loaded calibration data from %s\n", filename);
    return 0;
}

/* --------------------------------------------------------------------------
 * Main Application
 * ------------------------------------------------------------------------*/
 
/**
 * @brief Main function.
 *
 * Initializes the BNO055, handles calibration loading or execution,
 * and enters a loop to continuously read and display sensor data.
 *
 * @param argc Argument count.
 * @param argv Argument vector. Can accept "--calibrate" to force recalibration.
 * @return int EXIT_SUCCESS on success, EXIT_FAILURE on error.
 */
int main(int argc, char *argv[]) {
    int boot_mode = 0; // 0 for Normal, 1 for Forced Calibration
 
    if (argc > 1 && strcmp(argv[1], "--calibrate") == 0) {
        boot_mode = 1;
        printf("BOOT MODE: Forced Calibration\n");
    } else {
        printf("BOOT MODE: Normal\n");
    }
 
    // Verify sensor presence
    uint8_t chip_id = 0;
    if (smbus_read_byte_data(BNO055_I2C_BUS, BNO055_ADDRESS_A, BNO055_CHIP_ID_ADDR, &chip_id) != 0 || chip_id != 0xA0) {
        perror("BNO055 not found at I2C address 0x28. Check wiring.");
        return EXIT_FAILURE;
    }
    printf("BNO055 Chip ID: 0x%02X\n", chip_id);
 
    // Reset the sensor
    bno055_set_mode(BNO055_OPERATION_MODE_CONFIG);
    smbus_write_byte_data(BNO055_I2C_BUS, BNO055_ADDRESS_A, BNO055_SYS_TRIGGER_ADDR, 0x20);
    usleep(700000); // Wait for reset to complete
 
    // Load calibration or perform manual calibration
    if (boot_mode == 1 || bno055_load_calibration(CALIBRATION_FILENAME) != 0) {
        bno055_set_mode(BNO055_OPERATION_MODE_NDOF);
        usleep(25000);
        bno055_perform_manual_calibration();
        bno055_save_calibration(CALIBRATION_FILENAME);
    }
 
    // Set final operation mode for data reading
    printf("Setting final operation mode to NDOF...\n");
    bno055_set_mode(BNO055_OPERATION_MODE_NDOF);
    usleep(25000);
 
    // Main loop
    printf("\n--- Starting Continuous Data Readout ---\n");
    while(1) {
        read_all_sensor_data();
        usleep(100000); // 100ms delay, ~10Hz
    }
 
    return EXIT_SUCCESS;
}
 